source/mainnet/smart-contracts/tutorials/smartContractUpgrade/smartContractUpgrade.rst

```diff
@@ -0,0 +1,287 @@
+.. include:: ../../../variables.rst
+.. _smart-contract-upgrade:
+
+====================
+Native upgradability
+====================
+
+The goal of this tutorial is to deploy a ``contract-version1``, upgrade its logic and migrate its state to ``contract-version2``.
+
+The `contract-version1 <https://github.com/Concordium/concordium-rust-smart-contracts/tree/main/examples/smart-contract-upgrade/contract-version1>`_
+includes an ``upgrade`` function and the `contract-version2 <https://github.com/Concordium/concordium-rust-smart-contracts/tree/main/examples/smart-contract-upgrade/contract-version2>`_
+includes a ``migration`` function which the next two sections explore.
+
+``migration`` function
+======================
+
+A ``migration`` function can be optionally invoked as part of the upgrade to e.g. change the shape of the smart contract state.
+Use the below input file to specify the ``<EntryPointName>`` and the ``<InputParameters>``
+of the ``migration`` function.
+
+.. code-block:: json
+
+   {
+      "migrate": {
+         "Some": [
+            [
+            "<EntryPointName>",
+            "<InputParameters>"
+            ]
+         ]
+      },
+      "module": "<ModuleReferenceContractVersion2>"
+   }
+
+If the module reference should be upgraded without invoking a ``migration`` function, use the below input file instead:
+
+.. code-block:: json
+
+   {
+      "migrate": {
+         "None": [
+         ]
+      },
+      "module": "<ModuleReferenceContractVersion2>"
+   }
+
+``upgrade`` function
+====================
+
+To trigger a smart contract upgrade natively on Concordium, the upgrade needs to be triggered on the ``host`` as seen in the code snippet below:
+
+.. code-block:: rust
+
+   fn contract_name<S: HasStateApi>(
+      ...
+      host: &mut impl HasHost<S>,
+      ...
+   )  ... {
+      ...
+      let result = host.upgrade(module);
+      ...
+   }
+
+.. note::
+
+   If you don't find a ``host.upgrade(module)`` snippet in a smart contract code, you can be sure that this contract
+   can not be natively upgraded. The smart contract can still be part of a ``proxy-implementation`` pattern.
+   If you want to ensure a smart contract is truly immutable you need to do your own research
+   and study its logic.
+
+The function is marked as ``low_level``. This is ``necessary`` since the high-level mutable functions store the state of the contract at the end of
+execution. This conflicts with migration since the shape of the state ``might`` be changed by the ``migration`` function. If the state is then written
```

Comment:
```suggestion
execution. This conflicts with migration since the shape of the state *might* be changed by the ``migration`` function. If the state is then written
```